### ISSUE TITLE
Issue with bullet lists in `@param`

### ISSUE THREAD

**Issue by lionel-** [r-lib/roxygen2#1102]
Surfaced in r-lib/vctrs#1100.

This block:

```r
#' Foo
#' @name foo
#'
#' @param foo
#'   * A
#'   * B
#'   * C
NULL
```

produces a nested bullet list:

```Rd
\arguments{
\item{foo}{\itemize{
\item A
\itemize{
\item B
\item C
}
}}
}
```

The first workaround is to remove the indenting:

```r
#' Bar
#' @name bar
#'
#' @param bar
#' * A
#' * B
#' * C
NULL
```

```Rd
\arguments{
\item{bar}{\itemize{
\item A
\item B
\item C
}}
}
```

Or to add a header:

```r
#' Baz
#' @name baz
#'
#' @param baz
#'   Header.
#'   * A
#'   * B
#'   * C
NULL
```

```Rd
\arguments{
\item{baz}{Header.
\itemize{
\item A
\item B
\item C
}}
}
```

**Comment by gaborcsardi** [comment]
My output is slightly different.

``` r
roxygen2::roc_proc_text(roxygen2::rd_roclet(), "
  #' Foo
  #' @name foo
  #'
  #' @param foo
  #'   * A
  #'   * B
  #'   * C
  NULL
")
#> $foo.Rd
#> % Generated by roxygen2: do not edit by hand
#> % Please edit documentation in ./<text>
#> \name{foo}
#> \alias{foo}
#> \title{Foo}
#> \arguments{
#> \item{foo}{* A
#> * B
#> * C}
#> }
#> \description{
#> Foo
#> }
```

<sup>Created on 2020-05-28 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

**Comment by hadley** [comment]
We might be able to use the same algorithm as `glue::trim()` here

**Comment by gaborcsardi** [comment]
Since the white space is significant for markdown, maybe we should just not trim it at all? In some cases like this, at least. The specific problem here is that the markdown parser gets 
```
Browse[2]> text
[1] "* A\n  * B\n  * C"
```
and cutting off the initial white space here really matters.

**Comment by lionel-** [comment]
Since whitespace is significant, I would expect the roxy indenting to be cut off so it doesn't interfere with the markdown parser?

**Comment by lionel-** [comment]
Just to be clear, I meant that there are two kinds of whitespaces in a roxy markdown: (a) the indenting inside a roxy key, and (b) the whitespace for the markdown content. I would expect (a) to be removed so it doesn't interfere with (b). I understand that it might cause complications though.

> Then the rules are simple: we just cut off the #' comments (with white space before them), and the rest is treated as a markdown file.

These rules are simple but I think they produce unexpected behaviour in this case. Unindented roxy blocks are very hard to read, so we should encourage indenting.

**Comment by gaborcsardi** [comment]
> two kinds of whitespaces in a roxy markdown

How do you tell the difference? E.g. what kind of whitespace is this?

```r
#' @section Mysection:
#'     myfun()
#'     myfun2()
```

> but I think they produce unexpected behaviour in this case

To be clear, my suggest would solve your original problem.

**Comment by lionel-** [comment]
> E.g. what kind of whitespace is this?

I think using first indent as in glue to infer the indent space is reasonable and predictable. It would work both with `@param` and with `@description` sections. I don't see it causing problems, though of course I haven't thought much about these things.

> To be clear, my suggestion would solve your original problem.

Sorry I missed that. How does it solve it?

**Comment by gaborcsardi** [comment]
> I think using first indent as in glue to infer the indent space is reasonable and predictable.

Yes, but I think it is also not necessary. The best is not to have two-kinds of whitespace at all, just one kind. Much simpler. E.g. if you cut the whitespace, then you cannot easily move code between `@includeRmd` and inline comments, because you might need to adjust the white space.

> I don't see it causing problems, though of course I haven't thought much about these things.

The problem is that the formatting of this part of the example is context dependent:
```
#'     myfun()
#'     myfun2()
```

E.g. if you put this in an `.md` and include it, you get a code block. If we trim the white space then this is not the case for inline comments, because all the white space is removed.

We also do not know how white space will be significant in future versions of markdown, so just throwing it away does not seem like a good solution.

> Sorry I missed that. How does it solve it?

If we don't cut the whitespace on the first line, then the markdown parser will get this, and generate the correct output, i.e. a single list:

```
> cat(commonmark::markdown_xml("  * foo\n  * bar\n"))
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE document SYSTEM "CommonMark.dtd">
<document xmlns="http://commonmark.org/xml/1.0">
  <list type="bullet" tight="true">
    <item>
      <paragraph>
        <text xml:space="preserve">foo</text>
      </paragraph>
    </item>
    <item>
      <paragraph>
        <text xml:space="preserve">bar</text>
      </paragraph>
    </item>
  </list>
</document>
```

**Comment by lionel-** [comment]
> E.g. if you cut the whitespace, then you cannot easily move code between @includeRmd and inline comments, because you might need to adjust the white space.

This might be a feature. I think it is important to adjust the whitespace when copying from a .md to a roxy block because unindented roxy documentation is poor practice. Note that this only concerns a subset of keywords though, e.g. param and return, so other keywords that do not require indenting might skip the trimming.

> .g. if you put this in an .md and include it, you get a code block.

Is this the only expected problem? Fenced code blocks are much more common than indented blocks.

**Comment by gaborcsardi** [comment]
> unindented roxy documentation is poor practice

You can still indent it, in a way that works with markdown. E.g. your original post is perfectly fine. :)

> Or perhaps this is about whether the text starts on the same line as the keyword?

Yeah, this is exactly the problem with trimming. :) We would end up with a complicated set of rules that nobody understands. It would take a long time to explain them and make sure they are in sync with the docs, etc. Whereas not cutting is just that, not cutting. One line to explain.

> Is this the only expected problem? Fenced code blocks are much more common than indented blocks.

Indented blocks are canonical pandoc markdown, and the visual editor creates them as well. So they will be the standard in R, very likely.

I don't know if this is the only problem, or that it will be the only problem as markdown and commonmark evolves. But that's the point. If we don't process the text, then we don't need to know. Let the markdown parser interpret the markdown as it should.

**Comment by lionel-** [comment]
I think you may be overstating the complexity of these potential rules. People and IDEs will keep indenting their parameter lists and this whitespace sensitive syntax might be interpreted by the markdown parser in surprising ways. From this point of view, trimming the indent whitespace reduces the complexity of the syntax rather than increase it. I agree that finding a rule sufficiently simple and predictable to detect when it makes sense to trim might be difficult.

In any case we have the same goals, keeping things simple and having indentation that makes sense from the persectives of the roxygen and markdown syntaxes. I have found the markdown integration to work well in most cases, so if your suggestion solves the bullet list issue as well, I would be happy with it.

**Comment by nettoyoussef** [comment]
Related to this problem: listing parameter values inside an Rcpp function also doesn't work.

The function:
```cpp
//' Prints Hello World!
//'
//' @param my_param (int) - the level of verbosity. Select one of:
//' * 0 - does nothing
//' * 1 - still nothing
//' @export
// [[Rcpp::export]]
void helloworld(int my_param){
  Rprintf("Hello Word!");
}
```
gets exported to:

```R
% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/RcppExports.R
\name{helloworld}
\alias{helloworld}
\title{Prints Hello World!}
\usage{
helloworld(my_param)
}
\arguments{
\item{my_param}{(int) - the level of verbosity. Select one of:
* 0 - does nothing
* 1 - still nothing}
}
\description{
Prints Hello World!
}

```

**Comment by hadley** [comment]
It looks like the fix for this is depressingly simple.
